### PR TITLE
CDAP-765: access to custom and system metrics in unit-test framework

### PR DESCRIPTION
**Pull request by abaranau**

### CODE REVIEW

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT2394-1

**Comment by abaranau** [comment]
GREEN

**Comment by gsps1** [comment]
added a commit to for minor fixes. build running.

**Comment by gsps1** [comment]
http://builds.cask.co/browse/CDAP-DUT2394-2/test - build passed

File: cdap-client/src/main/java/co/cask/cdap/client/MetricsClient.java [hunk 1]
```diff
@@ -266,6 +266,16 @@ public void waitFor(String name, long count,
         doWaitFor(name, count, timeout, timeoutUnit);
       }
 
+      @Override
+      public long getCounter(String name) {
```
Comment (anwar6953) [inline cdap-client/src/main/java/co/cask/cdap/client/MetricsClient.java]: Why is the implementation the same for `getCounter` and `getGauge` in `MetricsClient`, but the implementations are different in `RuntimeStats`?

File: cdap-unit-test/src/main/java/co/cask/cdap/test/RuntimeStats.java [hunk 2]
```diff
@@ -182,6 +192,15 @@ public String toString() {
 
   private static long getTotalCounter(Map<String, String> context, String metricName) {
     MetricDataQuery query = getTotalCounterQuery(context, metricName);
+    return getSingleValueFromTotals(query);
+  }
+
+  private static long getLatestGauge(Map<String, String> context, String metricName) {
```
Comment (anwar6953) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/RuntimeStats.java]: actually, maybe the question is why do we need this `getLatestGauge` method with the same implementation as `getTotalCounter` above it?

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestFrameworkTestRun.java [hunk 3]
```diff
@@ -740,6 +740,35 @@ public void testAppWithServices() throws Exception {
...
   @Test
+  public void testWaitOnUserMetrics() throws Exception {
+
+    ApplicationManager applicationManager = deployApplication(AppWithUserMetrics.class);
+    applicationManager.getFlowManager("fooFlow").start();
+
+    // Send some inputs to streams
+    StreamManager streamManager = getStreamManager("fooStream");
+    for (int i = 0; i < 100; i++) {
+      streamManager.send(ImmutableMap.of("index", String.valueOf(i)), "testing message " + i);
```
Comment (anwar6953) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestFrameworkTestRun.java]: Do we need to send headers?
Doesn't seem like its used.

**Comment by anwar6953** [comment]
Also, it seems that the tags being queried are very limited (not customizable, like the metric name is).
Do you think that should also be more customizable? Perhaps at least in a separate PR or JIRA?

**Comment by gsps1** [comment]
@anwar6953 i don't think we need customization at tag level, as metrics are emitted from one of the program and we can get context for those programs, we are using those tags to get RuntimeMetrics.

**Comment by anwar6953** [comment]
The issue with RuntimeStats is that there are methods only for certain program types (flow, flowlet, service, service handler). There are many tags that are omitted (dataset, workflow, spark, mapreduce, etc).

**Comment by gsps1** [comment]
then we have to add implementation for those in their respective managers.

**Comment by anwar6953** [comment]
Why not simply expose a method like the `getMetrics` of `RuntimeStats` that allows user to pass in tags, and then query by metric name?

**Comment by gsps1** [comment]
Deprecated 
 1) RuntimeMetrics and 2) RuntimeStats classes. 

Using MetricsManager class, which has method to getMetrics<context, metric>  and getMetrics with retry. 

Updated TestFrameworkTestRun with dependent changes, using MetricsManager instead of deprecated RuntimeMetrics now.

Things remaining:
1) Change examples and tests using RuntimeMetrics to use MetricsManager
2) Change MetricsClient API to use MetricsManager instead of RuntimeMetrics and dependent changes.

will create JIRA to track the above two after the review and do it in a separate PR.

File: cdap-api/src/main/java/co/cask/cdap/api/metrics/RuntimeMetrics.java [hunk 4]
```diff
@@ -24,8 +24,8 @@
 /**
  * Provides access to metrics about running processors.
  *
- * TODO: Support custom metrics https://issues.cask.co/browse/CDAP-765
  */
+@Deprecated
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/metrics/RuntimeMetrics.java]: indicate what to use instead

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestFrameworkTestRun.java (hunk 3 above)
Comment (gsps1) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestFrameworkTestRun.java]: its used to emit, even/odd metrics in the flowlet.

File: cdap-unit-test/src/main/java/co/cask/cdap/test/MetricsManager.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  public Collection<MetricTimeSeries> query(MetricDataQuery query) throws Exception {
+    return metricStore.query(query);
+  }
+
+
+  /**
+   * waitForFlowlet's events processed metric, specified by count, and the timeout and timOutUnit is provided.
+   * @param namespace
```
Comment (anwar6953) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/MetricsManager.java]: Its not useful to have params in javadoc, if they're not going to have description, right?

File: cdap-api/src/main/java/co/cask/cdap/api/metrics/RuntimeMetrics.java [hunk 6]
```diff
@@ -23,9 +23,9 @@
 
 /**
  * Provides access to metrics about running processors.
- *
- * TODO: Support custom metrics https://issues.cask.co/browse/CDAP-765
+ * deprecated, Usages in MetricsClient and RuntimeStats will be replaced alternative implementation.
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/metrics/RuntimeMetrics.java]: I think it should be similar to deprecation description from other class:
`This is deprecated, should use {@link co.cask.cdap.test.MetricsManager} instead`

Is that description correct?

File: cdap-unit-test/src/main/java/co/cask/cdap/test/MetricsManager.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+                              Constants.Metrics.Tag.FLOW, flowName,
+                              Constants.Metrics.Tag.FLOWLET, flowletName
+    ), Constants.Metrics.Name.Flow.FLOWLET_PROCESSED, count, timeout, timeUnit);
+
+  }
+
+
+  /**
+   * waitFor a metric value count for the metric identified by metricName and context.
+   * @param context
```
Comment (anwar6953) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/MetricsManager.java]: same here

File: cdap-unit-test/src/main/java/co/cask/cdap/test/MetricsManager.java (hunk 5 above)
Comment (anwar6953) [inline cdap-unit-test/src/main/java/co/cask/cdap/test/MetricsManager.java]: app, flowName, flowletName are self-explanatory, but timeout and count are not (units?)

**Comment by gsps1** [comment]
removed deprecation for RuntimeMetrics, just deprecated RuntimeStats. using RuntimeMetrics in MetricsManager.

**Comment by anwar6953** [comment]
4 nit/trivial comments.
LGTM besides those.

**Comment by gsps1** [comment]
thanks @anwar6953 for the review. i have addressed your comments. please review again.

**Comment by gsps1** [comment]
build running : http://builds.cask.co/browse/CDAP-DUT2394-3

**Comment by gsps1** [comment]
build GREEN, merging

**Pull request merged**